src/main/java/de/azapps/kafkabackup/common/partition/cloud/S3PartitionWriter.java

```diff
@@ -1,76 +1,104 @@
-import java.io.ByteArrayOutputStream;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.connect.errors.RetriableException;
+
 import java.io.IOException;
-import lombok.Builder;
-import lombok.RequiredArgsConstructor;
+import java.util.LinkedList;
+import java.util.Queue;
 
-@RequiredArgsConstructor
-@Builder
+@Slf4j
 public class S3PartitionWriter implements PartitionWriter {
 
-  private final String bucketName;
-  private final int partition;
-
-  private final String topicName;
   private final AwsS3Service awsS3Service;
+  private final String bucketName;
+  private final TopicPartition topicPartition;
+  private final int maxBatchMessages;
+  private final long maxBatchTimeMs;
+  private final Queue<Record> buffer = new LinkedList<>();
 
-  private final RecordJSONSerde recordJSONSerde = new RecordJSONSerde();
+  private S3BatchWriter batchWriter;
+  private Long lastCommittableOffset = null; // Nothing written so far, so nothing to commit
 
-  private final long lastWrittenOffset = 0;
 
+  public S3PartitionWriter(AwsS3Service awsS3Service, String bucketName, TopicPartition tp, int maxBatchMessages, long maxBatchTimeMs) {
+    this.awsS3Service = awsS3Service;
+    this.bucketName = bucketName;
+    this.topicPartition = tp;
+    this.maxBatchMessages = maxBatchMessages;
+    this.maxBatchTimeMs = maxBatchTimeMs;
+  }
 
   @Override
   public void append(Record record) throws PartitionException {
-    String fileName = buildFileKeyForRecord(topicName, partition, record);
-    try {
-      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
-      recordJSONSerde.write(outputStream, record);
-      byte[] jsonRecord = outputStream.toByteArray();
-      ByteArrayInputStream jsonStream = new ByteArrayInputStream(jsonRecord);
+    buffer.add(record);
+  }
 
-      ObjectMetadata objectMetadata = new ObjectMetadata();
-      objectMetadata.setContentLength(jsonRecord.length);
+  @Override
+  public void flush() throws PartitionException, RetriableException {
+    while (!buffer.isEmpty()) {
+      Record record = buffer.poll();
+      writeToBatch(record);
+      maybeCommitBasedOnMessages();
+    }
+    maybeCommitBasedOnTime();
+  }
 
-      awsS3Service.saveFile(bucketName, fileName, jsonStream, objectMetadata);
+  private void writeToBatch(Record record) {
+    try {
+      if (batchWriter == null) {
+        batchWriter = new S3BatchWriter(awsS3Service, bucketName, topicPartition, record);
+      } else {
+        batchWriter.append(record);
+      }
     } catch (IOException e) {
-      e.printStackTrace();
+      throw new PartitionException(e);
     }
   }
 
-  private String buildFileKeyForRecord(String topic, int partition, Record record) {
-    return String.format("%s/%03d/msg_%020d.json", topic, partition, record.kafkaOffset());
+  private void maybeCommitBasedOnMessages() {
+    if (batchWriter != null && batchWriter.getCount() >= maxBatchMessages) {
```

Comment:
is case where batchWriter is null and this is "expected" or just "ok" state?  if not, shouldn't we throw an exception in this case?

Reply:
It's completely normal. The batch might already have been committed by some other mechanism (shouldn't assume any order between `maybeCommitBasedOnTime` and `maybeCommitBasedOnMessages`). And after a commit the batchWriter is `null` until we receive a new first message.